Ew2d eta corrections

ludwig-ahrens opened the pull request:
Corrections to the forces and energy from the electrodes.
This comprises three terms: Short range interaction, self interaction and potential energy (i.e. q * psi).
Only the first one leads to a correction of the forces.

Most of the setup() method of fix_charge_update is moved to setup_post_neighbour(). This way charges can be updated before the forces of the initial timestep are calculated.

ludwig-ahrens committed:
Use igroup to check if atoms in electrode

ludwig-ahrens committed:
Calculate eta terms of energy and force

ludwig-ahrens committed:
Charges, forces and energies before step 0

Most things in setup are moved to setup_post_neighbour, in order to
update the charges before the forces of the initial step are
calculated.
Eta corrections for the initial step are applied in setup().

robeme commented on src/USER-CONP/fix_charge_update.cpp:
```diff
@@ -35,12 +43,14 @@ FixChargeUpdate::FixChargeUpdate(LAMMPS *lmp, int narg, char **arg)
   f_inv = f_mat = f_vec = nullptr;
   read_inv = read_mat = false;
   symm = false;
+  scalar_flag = 1;
```
Das hast du gesetzt um die potentielle Energie zurückzugeben?

ludwig-ahrens replied:
Ja

robeme commented on src/USER-CONP/fix_charge_update.cpp:
```diff
@@ -253,6 +265,12 @@ void FixChargeUpdate::setup(int) {
+
+void FixChargeUpdate::setup(int) {
+  // correct forces for initial timestep
+  short_range_correction(true);
```
Ich würde das gaussian_charge_forces() nennen?

robeme replied:
gausscorr() wäre auch OK. Dann wäre man auch etwas mehr in der LAMMPS Lingo (siehe z.b. slabcorr).

ludwig-ahrens replied:
Habe es in gausscorr() umbenannt

robeme commented on src/USER-CONP/fix_charge_update.cpp:
```diff
@@ -361,11 +379,11 @@ void FixChargeUpdate::create_taglist() {
 
 std::vector<int> FixChargeUpdate::local_to_matrix() {
-  int const nmax = atom->nmax;
+  int const nall = atom->nlocal + atom->nghost;
```
nmax ist genau atom->nlocal + atom->nghost. Brauchst du also nicht so machen. Mit nmax würde es mehr der LAMMPS Lingo entsprechen.

ludwig-ahrens replied:
Nein, nmax ist größer als nall. Wenn ich über nmax iteriere bekomme ich Atome mit tag = 0, die aber trotzdem in igroup sind.

robeme commented on src/USER-CONP/fix_charge_update.cpp:
```diff
@@ -377,23 +395,144 @@ std::vector<int> FixChargeUpdate::local_to_matrix() {
+  int *mask = atom->mask;
   std::vector<int> mpos = local_to_matrix();
-  int const nmax = atom->nmax;
+  int const nall = atom->nlocal + atom->nghost;
```
ditto

robeme commented on src/USER-CONP/fix_charge_update.cpp:
```diff
@@ -377,23 +395,144 @@ std::vector<int> FixChargeUpdate::local_to_matrix() {
 /* ---------------------------------------------------------------------- */
 
+double FixChargeUpdate::compute_scalar() {
+  return electrode_energy(local_to_matrix()) + short_range_correction(false);
```
Für short_range_correction() vielleicht auch eine Überladung? Also ohne True gibt er einfach nur die Energien zurück?

ludwig-ahrens replied:
Ich finde es gut wenn man den Wert setzen muss. Dann vergisst man es auch nicht so leicht.
Können wir aber machen, wenn dir das wichtig ist.

robeme replied:
Vielleicht so machen wie in pair->single() mit den Kräften als übergebenen Pointer?

robeme commented on src/USER-CONP/fix_charge_update.cpp:
```diff
@@ -377,23 +395,144 @@ std::vector<int> FixChargeUpdate::local_to_matrix() {
+
+/* ---------------------------------------------------------------------- */
+
+double FixChargeUpdate::electrode_energy(vector<int> mpos) {
```
electrode_energy finde ich etwas zu generisch, wenn es eigentlich die self_interaction und die Energie durch das Potential Psi ist. Vielleicht aufspalten. Generell würde ich es gut finden, wenn wir die vier Beiträge self, kspace (bzw. LR), infinite boundary (bzw. keq0) sowie short (oder SR) immer getrennt nutzen und sie nicht irgendwie in anderen Methoden mischen – das verwirrt. Vielleicht können wir hier etwas konsistenter sein?

ludwig-ahrens replied:
Habe ich in potential_energy() und self_energy() aufgeteilt

robeme replied:
Was ist eigentlich genau die potential_energy()? Steht die irgendwo im MW Theo PDF?

ludwig-ahrens replied:
Habe ich da nicht gefunden, aber in coulomb_self.f90

robeme replied:
Ich würde das u.U. anders aufteilen. energy_self sollte direkt zur potentiellen Energie hinzugefügt werden (also zu eatom[i] bzw energy wie in ewald/conp.cpp 1235 if (eflag_global) energy += qscale * e_slabcorr;), weil das ist ja intrinisch in dem CONP Ansatz.

ludwig-ahrens committed:
Split and rename energy corrections

ludwig-ahrens committed:
Energy and force corrections in pre_reverse

Energy corrections via ev_tally are called in pre_reverse to avoid
reset in pair->compute after pre_force.
Total energy now includes potential energy q*psi.
compute_scalar includes q*psi, only.

robeme commented on src/USER-CONP/fix_charge_update.cpp:
```diff
@@ -555,6 +577,7 @@ int FixChargeUpdate::setmask() {
   int mask = 0;
   mask |= POST_NEIGHBOR;
   mask |= PRE_FORCE;
```
Brauchst du die Methode pre_force dann überhaupt noch? Ansonsten könnte die Methode aus der *.h und der *.cpp komplett entfernt werden und müsste konsequenterweise auch hier entfernt werden.

ludwig-ahrens replied:
pre_force brauche ich um die Ladungen neu zu setzen. Das muss vor pair->compute() etc. gemacht werden.
